graylog2-server/src/main/java/org/graylog2/outputs/OutputRegistry.java

```diff
@@ -179,8 +187,17 @@ public Callable<MessageOutput> loadForIdAndStream(final String id, final Stream
         return new Callable<>() {
             @Override
             public MessageOutput call() throws Exception {
-                final Output output = outputService.load(id);
-                return launchOutput(output, stream);
+                // Check if the output is still assigned to the given stream before loading and starting it.
+                // The stream assignment of the output could have been removed while the message object went
+                // through processing and output buffer processing handling.
+                // Without this check, we would start the output again after it has been stopped by removing it
+                // from a stream.
+                final Stream dbStream = streamService.load(stream.getId());
```

Comment:
I'm asking just for curiosity - have you checked if this additional communication with Mongo has some noticeable implications on performance?

Reply:
@luk-kaminski The database query to load the stream and the existing one below to load the output from the database are only executed if the output cannot be found in the cache. That means the query adds a bit of overhead when loading the cache entry.

When the server runs into the described race condition and message objects still have the output in their streams list, we might have a spike in database requests because of the additional query executed per message. But that only happens until all message objects created before removing the output from the stream are processed.

So yes, in these situations, there can be a performance hit. The only alternative I see to avoid the performance hit is to maintain an in-memory cache of all streams and their outputs to avoid the database calls. That requires careful cache management, though.

Reply:
Thanks for explanation, @bernd.
If performance decrease only happens in those rare situation, I vote with my both hands against any complications and for leaving the current solution! :)